clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -192,9 +179,9 @@ private static FilterResult filterTo(Iterable<MutableRecordBatch> batches, Recor
                     builder.overrideLastOffset(batch.lastOffset());
 
                 MemoryRecords records = builder.build();
-                destinationBuffer.position(destinationBuffer.position() + slice.position());
                 messagesRetained += retainedRecords.size();
                 bytesRetained += records.sizeInBytes();
+                outputBuffer = builder.buffer();
```

Comment:
If (1) outputBatch.size > inputBatch.size and (2) outputBatch.size > MaxMessageSize, perhaps we could log a warning that consumers before 0.10.2 may need to increase the fetch size?